Added string indexing

ammaralassal opened the pull request:
Implemented string indexing for Core.string
Handles references or struct values. No runtime checks as per https://discord.com/channels/655572317891461132/655578254970716160/1431015866270748682
Part of #6270

danakj commented on toolchain/diagnostics/diagnostic_kind.def:
```diff
@@ -397,6 +397,7 @@ CARBON_DIAGNOSTIC_KIND(AddrOnNonSelfParam)
 CARBON_DIAGNOSTIC_KIND(AddrOnNonPointerType)
 CARBON_DIAGNOSTIC_KIND(ArrayBoundTooLarge)
 CARBON_DIAGNOSTIC_KIND(ArrayBoundNegative)
+CARBON_DIAGNOSTIC_KIND(ArrayIndexNegative)
```
It feels like we could use `ArrayIndexOutOfBounds` for this as negatives are out of bounds?

danakj commented on core/prelude/types/string.carbon:
```diff
@@ -16,8 +23,15 @@ class String {
+    fn At[self: Self](subscript: i32) -> Char {
+      return StringAt(self, subscript);
+    }
+  }
```
```suggestion
  }

```

danakj replied:
There are some unresolved comments still, like this one

danakj commented on toolchain/check/eval.cpp:
```diff
@@ -1689,6 +1689,76 @@ static auto MakeConstantForBuiltinCall(EvalContext& eval_context,
 
     case SemIR::BuiltinFunctionKind::PrintChar:
     case SemIR::BuiltinFunctionKind::PrintInt:
+    case SemIR::BuiltinFunctionKind::StringAt:{
```
This has changed the behaviour of the other builtin functions that it's below in this switch, no?
Did you try running file_test?
Also, this change should come with some new tests (probably in check/ for compile-time use and in lower/ for run-time use).

danakj commented on toolchain/check/eval.cpp:
```diff
@@ -1689,6 +1689,76 @@ static auto MakeConstantForBuiltinCall(EvalContext& eval_context,
+      Phase phase = Phase::Concrete;
+      auto str_id = GetConstantValue(eval_context, arg_ids[0], &phase);
+      auto index_id = GetConstantValue(eval_context, arg_ids[1], &phase);
+      //If either isnt constant, we can't evaluate it at compile time
```
```suggestion
      // If either has no constant value, we can't evaluate it at compile time.
```

danakj commented on toolchain/check/eval.cpp:
```diff
@@ -1689,6 +1689,76 @@ static auto MakeConstantForBuiltinCall(EvalContext& eval_context,
+        eval_context.constant_values().GetInstId(ptr_const_id));
+      if (!string_literal){
+        return MakeNonConstantResult(phase);
+      }
```
Similarly, `GetAs`?

danakj commented on toolchain/check/eval.cpp:
```diff
@@ -1689,6 +1689,76 @@ static auto MakeConstantForBuiltinCall(EvalContext& eval_context,
+      
+      auto string_value = eval_context.sem_ir().string_literal_values().Get(
+        string_literal->string_literal_id);
+      //Get index value
```
```suggestion
```
This comment just says what the code says

danakj commented on toolchain/check/eval.cpp:
```diff
@@ -1689,6 +1689,76 @@ static auto MakeConstantForBuiltinCall(EvalContext& eval_context,
     case SemIR::BuiltinFunctionKind::PrintChar:
     case SemIR::BuiltinFunctionKind::PrintInt:
+    case SemIR::BuiltinFunctionKind::StringAt:{
+      //We need both the string and index to be constant
```
```suggestion
```
This is saying the same thing as the comment below?

danakj commented on toolchain/check/eval.cpp:
```diff
@@ -1689,6 +1689,76 @@ static auto MakeConstantForBuiltinCall(EvalContext& eval_context,
+      auto str_id = GetConstantValue(eval_context, arg_ids[0], &phase);
+      auto index_id = GetConstantValue(eval_context, arg_ids[1], &phase);
+      //If either isnt constant, we can't evaluate it at compile time
+      if (!str_id.has_value() || !index_id.has_value()) {
```
I don't think we need to support symbolic strings or indexes, so how about we also return here if phase is not Concrete?

Then we can stop with all the early outs below, and assume we have the concrete expected types, using GetAs instead of TryGetAs etc

danakj commented on toolchain/check/eval.cpp:
```diff
@@ -1689,6 +1689,76 @@ static auto MakeConstantForBuiltinCall(EvalContext& eval_context,
+      auto elements = eval_context.inst_blocks().Get(str_struct->elements_id);
+      if (elements.size() != 2) {
+        return MakeNonConstantResult(phase);
+      }
```
Similarly, we can CARBON_CHECK this.

danakj commented on toolchain/check/eval.cpp:
```diff
@@ -1689,6 +1689,76 @@ static auto MakeConstantForBuiltinCall(EvalContext& eval_context,
+      //Check for out of bounds
+      if (index_val.getActiveBits() > 64 ||
+          index_val.getZExtValue() >= string_value.size()) {
+        CARBON_DIAGNOSTIC(ArrayIndexOutOfBounds, Error,
```
A diagnostic name can only be used once, so maybe we want to add `StringAtIndexOutOfBounds`. Could we use that for the negative case too though?

We should add tests for both negative and too large as well.

danakj commented on toolchain/check/eval.cpp:
```diff
@@ -1689,6 +1689,76 @@ static auto MakeConstantForBuiltinCall(EvalContext& eval_context,
+
+      auto char_value = static_cast<uint8_t>(string_value[index_val.getZExtValue()]);
+
+      auto int_id = eval_context.ints().Add(llvm::APSInt(llvm::APInt(32, char_value), /*isUnsigned=*/false));
```
```suggestion
      auto int_id = eval_context.ints().Add(llvm::APSInt(llvm::APInt(8, char_value), /*isUnsigned=*/false));
```
I would expect an `8` here I think?

danakj commented on toolchain/lower/handle_call.cpp:
```diff
@@ -318,6 +318,45 @@ static auto HandleBuiltinCall(FunctionContext& context, SemIR::InstId inst_id,
     }
 
+    case SemIR::BuiltinFunctionKind::StringAt: {
+      // Get the string argument  
```
```suggestion
```
This just says what the code says

danakj commented on toolchain/lower/handle_call.cpp:
```diff
@@ -318,6 +318,45 @@ static auto HandleBuiltinCall(FunctionContext& context, SemIR::InstId inst_id,
+      auto string_inst_id = arg_ids[0];
+      auto* string_arg = context.GetValue(string_inst_id);
+      
+      // Check if this is a pointer to a String or a String value
```
```suggestion
      // Check if this is a pointer to a String or a String value.
```
All comments should be complete sentences, I won't keep commenting on them all here

danakj commented on toolchain/sem_ir/builtin_function_kind.cpp:
```diff
@@ -322,6 +322,10 @@ constexpr BuiltinInfo PrintInt = {
 
+// Gets a character from a string at the given index.
+constexpr BuiltinInfo StringAt = {
+    "string.at", ValidateSignature<auto(AnyType, AnySizedInt)->AnySizedInt>};
```
Can we validate that the first parameter is a ClassType of Core.String? And we could write a test with an incorrect type that fails?

ammaralassal replied:
I will update the signature validation to require the first parameter to be a ClassType of Core.String instead of AnyType. I can also add a test that attempts to declare StringAt with an incorrect first parameter type. Regarding the test, is there an existing pattern I should follow for validating class types in built-in signatures?

danakj replied:
The `Check` method has a `TypeId`, which has the information you need. The type instruction would be a [`ClassType`](https://github.com/carbon-language/carbon-lang/blob/1ece5000aabfbe553fcd844b28e6bfedd6b3d181/toolchain/sem_ir/typed_insts.h#L453-L463), which has a [`ClassId`](https://github.com/carbon-language/carbon-lang/blob/1ece5000aabfbe553fcd844b28e6bfedd6b3d181/toolchain/sem_ir/ids.h#L317) that is [pointing to](https://github.com/carbon-language/carbon-lang/blob/1ece5000aabfbe553fcd844b28e6bfedd6b3d181/toolchain/check/context.h#L265) a [`Class`](https://github.com/carbon-language/carbon-lang/blob/1ece5000aabfbe553fcd844b28e6bfedd6b3d181/toolchain/sem_ir/class.h#L15-L90).

danakj commented on toolchain/lower/handle_call.cpp:
```diff
@@ -318,6 +318,45 @@ static auto HandleBuiltinCall(FunctionContext& context, SemIR::InstId inst_id,
+      } else {
+        // Already have the struct value
+        string_value = string_arg;
+      }
```
Eval expects the argument to be a StructValue, why are we checking for pointers here?

ammaralassal replied:
I was not sure whether the String argument would be passed by value or by reference at this point in lowering. Is there a helper I should use to determine the representation? Instead of checking the LLVM type?

danakj replied:
Hm I am not very familiar lower right now, but it should at least be consistent. It's being received as a "value", and classes have a pointer value-representation by default, so maybe this is always a pointer. Running the test should show you what happens, and I would expect it to be consistent.

danakj commented on toolchain/lower/handle_call.cpp:
```diff
@@ -318,6 +318,45 @@ static auto HandleBuiltinCall(FunctionContext& context, SemIR::InstId inst_id,
+      
+      // Extend to i32 
+      context.SetLocal(inst_id, context.builder().CreateZExt(
+          char_i8, context.GetTypeOfInst(inst_id), "string.char.zext"));
```
Is this extending to i32? Isn't it going to whatever the return type was? Maybe drop the comment?

danakj reviewed (commented):
It looks like you might want to set up pre-commit so you can check that it will pass that before uploading: https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/contribution_tools.md#running-pre-commit

ammaralassal committed:
Address feedback

ammaralassal commented:
I have addressed most of the feedback. 
* There is now a constraint in place `CoreStringType` that validates the parameter of `StringAt` to be of type `Core.String`.
* >I don't think we need to support symbolic strings or indexes, so how about we also return here if phase is not Concrete?
* `StringAt` was in the wrong case block, an oversight on my part.
* No longer checks if the String is a pointer as it always was a pointer.
* Added a test case that validates the use of StringAt with an incorrect parameter type.

Another issue that I came across was somehow being able to use `Core.String` in tests. I was not able to get that to work. On that note, if there is a way for me to do so, I can add tests for negative/out of bounds indices.

josh11b commented:
> Another issue that I came across was somehow being able to use `Core.String` in tests. I was not able to get that to work. On that note, if there is a way for me to do so, I can add tests for negative/out of bounds indices.

I think string literals should be able to convert to `Core.String` (I believe `str` is another name for `Core.String`). See for example: https://github.com/search?q=repo%3Acarbon-language%2Fcarbon-lang+path%3A%2F%5Etoolchain%5C%2Fcheck%5C%2Ftestdata%5C%2F%2F+%2F%28%3F-i%29%3A+str%5Cb%2F&type=code

ammaralassal committed:
Addressed feedback further.

ammaralassal committed:
Removed unused library

ammaralassal commented:
* Moved bounds checking to when the `[ ]` operator is used with strings at compile-time.
Tests are consistent now (Thank you all for your help on that!).

ammaralassal committed:
Duplicate comment

ammaralassal committed:
More tests

ammaralassal commented:
I think this is ready to be reviewed again if possible

danakj commented:
Note, please try to avoid rebase/force push in between review iterations, as it removes the ability to see "changes since the last review" and detaches comments from their places.

danakj commented on toolchain/check/testdata/operators/overloaded/string_indexing.carbon:
```diff
@@ -0,0 +1,26 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// INCLUDE-FILE: toolchain/testing/testdata/min_prelude/full.carbon
+// AUTOUPDATE
+// TIP: To test this file alone, run:
+// TIP:   bazel test //toolchain/testing:file_test --test_arg=--file_tests=toolchain/check/testdata/operators/overloaded/string_indexing.carbon
+// TIP: To dump output, run:
+// TIP:   bazel run //toolchain/testing:file_test -- --dump_output --file_tests=toolchain/check/testdata/operators/overloaded/string_indexing.carbon
+
+
+// --- test_string_indexing.carbon
+
+
```
```suggestion
library "[[@TEST_NAME]]";

```

danakj commented on toolchain/check/testdata/operators/overloaded/string_indexing.carbon:
```diff
@@ -0,0 +1,26 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// INCLUDE-FILE: toolchain/testing/testdata/min_prelude/full.carbon
```
```suggestion
//
// INCLUDE-FILE: toolchain/testing/testdata/min_prelude/full.carbon
//
```

danakj commented on toolchain/check/testdata/operators/overloaded/string_indexing.carbon:
```diff
@@ -0,0 +1,26 @@
+
+fn PrintStr(msg: str) {
+  for (i: i32 in Core.Range(msg.Size() as i32)) {
+    Core.PrintChar(msg[i]);
```
Do we need to print things to test indexing? I think we could just make a string, and then do an index into it and wrap that in `//@dump-sem-ir-begin` and `//@dump-sem-ir-end` so that we can see what semir comes out of it?

danakj commented on toolchain/check/testdata/operators/overloaded/string_indexing.carbon:
```diff
@@ -0,0 +1,26 @@
+
+fn Run() {
+  PrintStr("Hello World!\n");
+}
```
no Run in tests

danakj commented on toolchain/check/testdata/operators/overloaded/string_indexing_negative.carbon:
```diff
@@ -0,0 +1,21 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// INCLUDE-FILE: toolchain/testing/testdata/min_prelude/full.carbon
```
```suggestion
//
// INCLUDE-FILE: toolchain/testing/testdata/min_prelude/full.carbon
//
```

danakj commented on toolchain/check/testdata/operators/overloaded/string_indexing_negative.carbon:
```diff
@@ -0,0 +1,21 @@
+
+
+// --- fail_negative_index.carbon
+
```
```suggestion
library "[[@TEST_NAME]]";

```

danakj commented on toolchain/check/testdata/operators/overloaded/string_indexing_negative.carbon:
```diff
@@ -0,0 +1,21 @@
+// TIP:   bazel run //toolchain/testing:file_test -- --dump_output --file_tests=toolchain/check/testdata/operators/overloaded/string_indexing_negative.carbon
+
+
+// --- fail_negative_index.carbon
```
Consider moving this test into the string_indexing.carbon file, so we only have to compile the prelude once?

danakj commented on toolchain/check/testdata/operators/overloaded/string_indexing_negative.carbon:
```diff
@@ -0,0 +1,21 @@
+  // CHECK:STDERR:   let c: Core.Char = test_str[-1];
+  // CHECK:STDERR:                               ^~
+  // CHECK:STDERR:
+  let c: Core.Char = test_str[-1];
```
```suggestion
  test_str[-1];
```

danakj commented on toolchain/check/testdata/operators/overloaded/string_indexing_negative.carbon:
```diff
@@ -0,0 +1,21 @@
+// --- fail_negative_index.carbon
+
+fn TestNegativeIndex() {
+  let test_str: str = "Test";
```
```suggestion
  let s: str = "Test";
```

Everything is a test in tests :) I would just go with a simple name

danakj commented on toolchain/check/testdata/operators/overloaded/string_indexing_out_of_bounds.carbon:
```diff
@@ -0,0 +1,17 @@
+// TIP: To dump output, run:
+// TIP:   bazel run //toolchain/testing:file_test -- --dump_output --file_tests=toolchain/check/testdata/operators/overloaded/string_indexing_out_of_bounds.carbon
+
+// --- fail_out_of_bounds.carbon
```
Move this into the string_indexing.carbon file?

danakj commented on toolchain/check/testdata/operators/overloaded/string_indexing_out_of_bounds.carbon:
```diff
@@ -0,0 +1,17 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// INCLUDE-FILE: toolchain/testing/testdata/min_prelude/full.carbon
```
```suggestion
//
// INCLUDE-FILE: toolchain/testing/testdata/min_prelude/full.carbon
//
```

danakj commented on toolchain/check/testdata/operators/overloaded/string_indexing_out_of_bounds.carbon:
```diff
@@ -0,0 +1,17 @@
+// CHECK:STDERR: var c: Core.Char = "Test"[4];
+// CHECK:STDERR:                           ^
+// CHECK:STDERR:
+var c: Core.Char = "Test"[4];
```
```suggestion
var c: char = "Test"[4];
```

danakj replied:
In these 3 tests we have 3 pretty different formulations of the test structure, it'd be nice to just have one simple structure that we repeat. I think just

```
fn F() {
  let s: str = "Stuff";
  //@dump-sem-ir-begin
  ... test condition here ...
  //@dump-sem-ir-end
}
```
Would be simple and clear?

danakj commented on toolchain/check/testdata/operators/overloaded/string_indexing_wrong_type.carbon:
```diff
@@ -0,0 +1,20 @@
+// TIP: To dump output, run:
+// TIP:   bazel run //toolchain/testing:file_test -- --dump_output --file_tests=toolchain/check/testdata/operators/overloaded/string_indexing_wrong_type.carbon
+
+// --- fail_wrong_type.carbon
```
Move into string_indexing.carbon?

danakj commented on toolchain/check/eval.cpp:
```diff
@@ -1689,6 +1689,76 @@ static auto MakeConstantForBuiltinCall(EvalContext& eval_context,
+        SemIR::IntValue{.type_id = call.type_id, .int_id = int_id},
+        phase);
+
+    }
```
This suggestion looks weird, does it no longer apply? Let's resolve it if so.

danakj commented on toolchain/check/eval.cpp:
```diff
@@ -1687,6 +1687,39 @@ static auto MakeConstantForBuiltinCall(EvalContext& eval_context,
+
+      auto ptr_const_id = eval_context.constant_values().Get(elements[0]);
+      auto string_literal = eval_context.insts().GetAs<SemIR::StringLiteral>(
+          eval_context.constant_values().GetInstId(ptr_const_id));
```
This probably will need to change when we have ways to construct a compile-time string from another compile-time string, but I think it's okay for now.

danakj commented on toolchain/check/eval.cpp:
```diff
@@ -1687,6 +1687,39 @@ static auto MakeConstantForBuiltinCall(EvalContext& eval_context,
+      auto string_literal = eval_context.insts().GetAs<SemIR::StringLiteral>(
+          eval_context.constant_values().GetInstId(ptr_const_id));
+
+      auto string_value = eval_context.sem_ir().string_literal_values().Get(
```
```suggestion
      const auto& string_value = eval_context.sem_ir().string_literal_values().Get(
```
Avoid copying the string

danakj commented on toolchain/check/testdata/operators/overloaded/string_indexing_wrong_type.carbon:
```diff
@@ -0,0 +1,20 @@
+  // CHECK:STDERR:   var c: Core.Char = x[0];
+  // CHECK:STDERR:                      ^~~~
+  // CHECK:STDERR:
+  var c: Core.Char = x[0];
```
```suggestion
  var c: char = x[0];
```

danakj commented on toolchain/check/handle_index.cpp:
```diff
@@ -26,6 +26,68 @@ auto HandleParseNode(Context& /*context*/, Parse::IndexExprStartId /*node_id*/)
+                                   SemIR::InstId index_inst_id,
+                                   const llvm::APInt& index_int) -> void {
+  if (index_int.isNegative()) {
+    CARBON_DIAGNOSTIC(ArrayIndexNegative, Error, "index `{0}` is negative.",
```
```suggestion
    CARBON_DIAGNOSTIC(StringIndexNegative, Error, "index `{0}` is negative.",
```
We can't reuse the same diagnostic in more than one place.

danakj replied:
```suggestion
    CARBON_DIAGNOSTIC(StringIndexNegative, Error, "index `{0}` is negative.",
```

danakj commented on toolchain/check/handle_index.cpp:
```diff
@@ -26,6 +26,68 @@ auto HandleParseNode(Context& /*context*/, Parse::IndexExprStartId /*node_id*/)
+  if (index_int.getActiveBits() > 64 ||
+      index_int.getZExtValue() >= string_value.size()) {
+    CARBON_DIAGNOSTIC(StringAtIndexOutOfBounds, Error,
+                      "string index `{0}` is past the end of the string.",
```
How about something that says the bound in it?
```suggestion
                      "string index `{0}`is out of bounds; string has length {1}.",
```

danakj commented on toolchain/check/handle_index.cpp:
```diff
@@ -84,6 +146,27 @@ auto HandleParseNode(Context& context, Parse::IndexExprId node_id) -> bool {
+    case CARBON_KIND(SemIR::ClassType class_type): {
+      if (IsStringType(context, class_type)) {
+        auto index_const_id = context.constant_values().Get(index_inst_id);
+        if (index_const_id.is_constant()) {
```
Given these bounds checks are only done for constant values, I am wondering why they are happening here instead of inside eval, more closely matching `ArrayIndex` behaviour? I think then we don't need the IsStringType check in here at all?

ammaralassal replied:
The bounds checks happen here because string indexing goes through the `IndexWith` interface, not primitive array indexing. I thought to do this to catch constant index errors earlier during semantic analysis before constant evaluation.
I noticed this during tests as well that when I was checking bounds that it only threw errors for fully constant expressions `"Test"[4]`, but it failed to throw for variable bindings `let s: str = "Test"; let c: char = s[4];` Even though both the string value and index are compile-time constants. By checking here, the implementation can check bounds for both direct literal indexing and variable-bound strings with constant indices.

danakj replied:
I fear I was not clear. I am not wondering why it's using the ArrayIndex implementation. I am wondering if the checking can be done in the implementation of StringAt in eval: https://github.com/carbon-language/carbon-lang/pull/6329/files/153ff8269fd4a57553db0ec1a000d98ebb3279a2#diff-89653437c0b9dd4925413361e1a840632f5f4dd03dc63fee83ce2d04a1586bf1R1716 we have all the information we need here to check and return an Error instead of duplicating a lot of the code here.

ammaralassal replied:
I attempted to move the bounds checking to eval, but the `StringAt` builtin is never being invoked during the check phase for non-constant strings.

```
// --- fail_negative_index.carbon
library "[[@TEST_NAME]]";

fn TestNegativeIndex() {
  let s: str = "Test";
  //@dump-sem-ir-begin
  let c: char = s[-1];
  //@dump-sem-ir-end
}
```
For example, the SemIR dump from this test case
`%str.as.IndexWith.impl.At.call: init %char = call %bound_method.loc6_21.2(%s.ref, %.loc6_19.2)`

This call does not have the the `[concrete]` annotation, implying it is not being evaluated as a constant.

I suppose the issue lies in `MakeConstantForCall`, more specifically
```
 // TODO: Some builtin calls might allow some operands to be non-constant.
  if (!has_constant_operands) {
    if (builtin_kind.IsCompTimeOnly(
            eval_context.sem_ir(), eval_context.inst_blocks().Get(call.args_id),
            call.type_id)) {
      CARBON_DIAGNOSTIC(NonConstantCallToCompTimeOnlyFunction, Error,
                        "non-constant call to compile-time-only function");
      CARBON_DIAGNOSTIC(CompTimeOnlyFunctionHere, Note,
                        "compile-time-only function declared here");
      const auto& function = eval_context.functions().Get(
          std::get<SemIR::CalleeFunction>(callee).function_id);
      eval_context.emitter()
          .Build(inst_id, NonConstantCallToCompTimeOnlyFunction)
          .Note(function.latest_decl_id(), CompTimeOnlyFunctionHere)
          .Emit();
    }
    return SemIR::ConstantId::NotConstant;
  }
  ```
  Since the string variable `s` in my test is not constant, the call returns `NotConstant` before ever reaching `MakeConstantForBuiltinCall` where the `StringAt` builtin is. I suppose this limitation is known based on the comment in `MakeConstantForCall`, however would it be worthwhile to modify `MakeConstantForCall` to allow `StringAt` (and potentially other builtins) to be invoked even when operands are not constant?

danakj replied:
Eval is constant evaluation. Each instruction has a constant value that is either constant (an instruction representing the constant result) or that is "runtime" (aka NotConstant). See `ConstantId::is_constant`.

If all the operands of (inputs to) the instruction are constant, then the resulting value can also be constant. That's when we execute StringAt in eval, and produce a constant value as a result. And since we have constant inputs, we can do bounds checks there.

If any of the operands of the instruction are runtime, as in the example you wrote here, then we can not evaluate to a constant value during compile (in eval). Instead, we the call instruction goes through to lower, and we generate runtime code for it.

Does that make sense?

ammaralassal replied:
That makes sense yes, sorry for misunderstanding! So with my current implementation I have bounds checking only during semantic analysis `handle_index`. And the actual character extraction without bounds checking in `eval`, with `StringAt` in lower to handle runtime code generation. In practice, I have found that if bounds checking were only done when evaluating a call to `StringAt` in eval, it would not catch all the cases we want. Specifically, cases with variable bindings because the call instruction does not get evaluated as a constant, it is treated as a runtime call even though both operands are constant values. Looking at `ConstantId::is_constant`, the issue appears to be that a NameRef to a `let` binding may itself have a `ConstantId` of `NotConstant`, even though the bindings value is constant. A call is evaluated as a constant if all operands are constant, and this check would happen before `GetConstantValue` can resolve. By only checking in `handle_index`, we can perform the bounds checking during semantic analysis by resolving through `NameRef` / `ValueBinding` to find constant string values, catching errors earlier than relying on constant evaluation of the Call instruction. Is this approach sufficient?

danakj replied:
> a NameRef to a let binding may itself have a ConstantId of NotConstant, even though the bindings value is constant

That sounds surprising. The implementation of eval for NameRef is this: https://github.com/carbon-language/carbon-lang/blob/315b0ac2410f8bd07ea80c9f9789663255f6d829/toolchain/check/eval_inst.cpp#L478-L483

Which just uses the constant value of the `value_id` operand.

Do you have a minimal example, which we can look at the semir for?

Here's an example of a `NameRef(X)` with a constant value: https://carbon.godbolt.org/z/G36GdPPrT

ammaralassal replied:
Using the following test case
```
// --- test_string_indexing.carbon
library "[[@TEST_NAME]]";

fn TestStringIndexing() {
  let s: str = "Test";
  //@dump-sem-ir-begin
  letc: char = s[0];
  //@dump-sem-ir-end
}
```

The semir shows:
```
%s.ref: ref %str.ee0ae5.1 = name_ref s, %s
%int_0: Core.IntLiteral = int_value 0 [concrete = constants.%int_0]
...
%str.as.IndexWith.impl.At.call: init %char = call %bound_method.loc6_20.2(%.loc6_17, %int_0)
```

The NameRef `%s.ref` does not have the `[concrete]` annotation, and neither does the resulting Call instruction. The index
is marked `[concrete]`, but the call is not being evaluated as constant.

Even trying a more literal case, where the string itself is concrete
```
// --- fail_literal_negative_index.carbon
library "[[@TEST_NAME]]";

fn TestLiteralNegativeIndex() {
  //@dump-sem-ir-begin
  let c: char = "Test"[-1];
  //@dump-sem-ir-end
}
```
Generates this semir
```
%str: %ptr.fb0 = string_literal "Test" [concrete = constants.%str.0a6]
...
%String.val: %str.ee0ae5.1 = struct_value (%str, %int_4) [concrete = constants.%String.val]
...
str.as.IndexWith.impl.At.call: init %char = call %bound_method.loc9_26.2(%String.val, %.loc9_24.2)
```
The call itself still is not being evaluated as constant. Is there something preventing interface method calls from being constant evaluated?

danakj replied:
In your first example:
```
  let s: str = "Test";
```
This is a runtime value. a `:` denotes a runtime binding.

>  Is there something preventing interface method calls from being constant evaluated?

Here's another example of an interface with builtin methods: https://github.com/carbon-language/carbon-lang/blob/315b0ac2410f8bd07ea80c9f9789663255f6d829/core/prelude/types/int.carbon#L52-L55

Here is us calling that to generate a compile-time constant value: https://carbon.godbolt.org/z/jnfTcc4ee (the semir gets truncated)

danakj replied:
I think this might help:

The `[concrete  =` on the RHS of the textual semir is the constant value that was output from eval. The call instruction is going into eval, and it is not returning a concrete constant value. It looks like it's returning `ConstantId::NotConstant`, and you'll probably want to run things in a debugger to figure out what exactly is going on.

ammaralassal replied:
Heres what I found, note output from lldb

```
TEST: toolchain/check/testdata/operators/overloaded/string_indexing.carbon (lldb)  p builtin_kind
(Carbon::SemIR::BuiltinFunctionKind) $0 = {
  Carbon::Internal::EnumBase<Carbon::SemIR::BuiltinFunctionKind, Carbon::SemIR::Internal::BuiltinFunctionKindData::RawEnum, const llvm::StringLiteral *> = (value_ = None)
}
```

The test attempts to evaluate `s[0]`, which would resolve to the current implementation of `At` within the `IndexWith` interface. However, when the evaluator reaches eval, we hit this check
```
if (builtin_kind == SemIR::BuiltinFunctionKind::None) {
      // TODO: Eventually we'll want to treat some kinds of non-builtin
      // functions as producing constants.
      return SemIR::ConstantId::NotConstant;
    }
```
`At` is not a builtin,  so we get back `NotConstant` thereby not hitting `StringAt` all.

Is there a way to make the constant evaluator "see through" the `At` wrapper to reach the `StringAt` builtin?

(Apologies for all the questions on this PR I really appreciate your patience!)

danakj replied:
I see, right. Note the different way that EqWith is written:
```carbon
    fn At[self: Self](subscript: T) -> Char {
      return StringAt(self, subscript);
    }
```
```carbon
  fn Equal[self: Self](other: UInt(M)) -> bool = "int.eq";
```

Instead of writing a function that _calls_ a builtin internally, we should make the At function into the builtin. Sorry I overlooked that this whole time, I didn't forsee that issue.

ammaralassal replied:
I made some progress, literal cases are evaluated correctly
```
fn TestLiteralOutOfBounds() {
  //@dump-sem-ir-begin
  // CHECK:STDERR: fail_literal_out_of_bounds.carbon:[[@LINE+4]]:17: error: string index `4` is out of bounds; string has length 4. [StringAtIndexOutOfBounds]
  // CHECK:STDERR:   let c: char = "Test"[4];
  // CHECK:STDERR:                 ^~~~~~~~~
  // CHECK:STDERR:
  let c: char = "Test"[4];
  //@dump-sem-ir-end
}
```

However, cases with variable bindings fail to throw. Would it be acceptable to add a special case for `StringAt` in `MakeConstantForCall`? Currently, `MakeConstantForCall` has the following check
```
 // If any operand of the call is non-constant, the call is non-constant.
  if (!has_constant_operands){
  ...
  }
```
  I also implemented a method that traces through `NameRef`, `ValueBinding` and `value_id` to access the constant initializer, however it is only invoked if we allow `StringAt` to process despite having "non-constant" operands. 

 More specifically
 ```
 // If any operand of the call is non-constant, the call is non-constant.
  if (!has_constant_operands) {
    if (builtin_kind == SemIR::BuiltinFunctionKind::StringAt) {
      return MakeConstantForBuiltinCall(
          eval_context, SemIR::LocId(inst_id), call, builtin_kind,
          eval_context.inst_blocks().Get(call.args_id), phase);
    }
```

danakj replied:
That would do the wrong thing here:
```
fn F(s: str) {
  s[2];
}
```
The `s` is a runtime value, we can not do a compile time indexing operation on it.

So I'm not sure what you're trying to address, is it this?
```
fn F(s:! str) {
  s[2];
}
```
The `s` there has a constant value, though it is symbolic. I would expect that to get past `MakeConstantForCall`. We have other compile time functions that work with symbolics. So I am not confident if this is the scenario...

But then I would expect that to end up failing inside the compile-time implementation of StringAt when it can't find a *concrete* constant value. And it's okay for us to not handle symbolic values for now, to do so we'd need to introduce a symbolic evaluation result (some new instruction) which evaluates to the result of string indexing once a more specific string value was known.

But that `MakeConstantForCall` is involved makes me uncertain about what is failing here for you/what you're trying to make work.

danakj commented on toolchain/sem_ir/builtin_function_kind.cpp:
```diff
@@ -177,6 +177,22 @@ struct AnyType {
     return true;
   }
 };
+// Constraint that checks if a type is Core.String.
```
```suggestion

// Constraint that checks if a type is Core.String.
```

danakj commented on toolchain/sem_ir/builtin_function_kind.cpp:
```diff
@@ -177,6 +177,22 @@ struct AnyType {
+    auto type_inst_id = sem_ir.types().GetInstId(type_id);
+    auto class_type = sem_ir.insts().TryGetAs<ClassType>(type_inst_id);
+    if (!class_type) {
+      // Not a string.
```
```suggestion
```

danakj commented on toolchain/sem_ir/builtin_function_kind.cpp:
```diff
@@ -177,6 +177,22 @@ struct AnyType {
+    }
+
+    const auto& class_info = sem_ir.classes().Get(class_type->class_id);
+
```
```suggestion
```

ammaralassal committed:
Addressed more feedback.

ammaralassal commented:
- The `At` method in the `IndexWith` impl is now the builtin 
 - Changed `StringAt` signature to allow for `AnyType` to support generic type parameters
 - ~~`ValueBinding` now propagates constant values for compile-time bounds checking for strings with variable bindings~~

ammaralassal committed:
Reverted some changes from previous commit

ammaralassal commented:
I have addressed all the feedback.
I apologize, this whole time I was trying to also have this work for symbolic values. I believe this can be reviewed again. 
Thank you!

danakj commented on core/prelude/types/string.carbon:
```diff
@@ -8,16 +8,23 @@ import library "prelude/copy";
+import library "prelude/operators/as";
 
 class String {
+
```
```suggestion
```

danakj commented on core/prelude/types/string.carbon:
```diff
@@ -8,16 +8,23 @@ import library "prelude/copy";
 import library "prelude/types/uint";
+import library "prelude/operators/index";
+import library "prelude/types/int";
+import library "prelude/operators/as";
```
Is this needed?

ammaralassal replied:
Yes, `.../index` provides the `IndexWith` interface since `String` implements it, `.../int` provides the `i64` type as `IndexWith` requires `ImplicitAs(i64)` and `.../as` provides the `ImplicitAs` interface as the generic parameter `T` is constrained with it.

danakj commented on core/prelude/types/string.carbon:
```diff
@@ -8,16 +8,23 @@ import library "prelude/copy";
   private var size: u64;
 }
+
+impl forall [T:! ImplicitAs(i32)] String as IndexWith(T) where .ElementType = Char {
```
I think we want to allow strings larger than what i32 can index.

```suggestion
impl forall [T:! ImplicitAs(i64)] String as IndexWith(T) where .ElementType = Char {
```

danakj commented on toolchain/check/eval.cpp:
```diff
@@ -1687,6 +1687,64 @@ static auto MakeConstantForBuiltinCall(EvalContext& eval_context,
+
+      auto ptr_const_id = eval_context.constant_values().Get(elements[0]);
+      auto string_literal = eval_context.insts().GetAs<SemIR::StringLiteral>(
+          eval_context.constant_values().GetInstId(ptr_const_id));
```
I think I'd combine these
```suggestion
      auto string_literal = eval_context.insts().GetAs<SemIR::StringLiteral>(
          eval_context.constant_values().GetConstantInstId(elements[0]));
```

danakj commented on toolchain/check/eval.cpp:
```diff
@@ -1687,6 +1687,64 @@ static auto MakeConstantForBuiltinCall(EvalContext& eval_context,
+
+      auto str_struct = eval_context.insts().GetAs<SemIR::StructValue>(str_id);
+      auto elements = eval_context.inst_blocks().Get(str_struct.elements_id);
+      CARBON_CHECK(elements.size() == 2, "String struct should have 2 fields.");
```
It's not obvious what the second element is, and we don't use it here. Can we leave a comment saying what the two fields are?

danakj commented on toolchain/check/eval.cpp:
```diff
@@ -1687,6 +1687,64 @@ static auto MakeConstantForBuiltinCall(EvalContext& eval_context,
+      }
+
+      if (index_val.getActiveBits() > 64 ||
+          index_val.getZExtValue() >= string_value.size()) {
```
```suggestion
      if (index_val.getZExtValue() >= string_value.size()) {
```
I think we only need this, right? Or is there something loadbearing about the 64 check?

danakj commented on toolchain/check/eval.cpp:
```diff
@@ -1687,6 +1687,64 @@ static auto MakeConstantForBuiltinCall(EvalContext& eval_context,
+            loc_id, StringAtIndexOutOfBounds,
+            {.type = eval_context.insts().Get(arg_ids[1]).type_id(),
+             .value = index_val},
+            static_cast<unsigned>(string_value.size()));
```
This is a narrowing conversion from size_t to unsigned. Should we just be emitting it as `size_t`?

danakj commented on toolchain/check/eval.cpp:
```diff
@@ -1687,6 +1687,64 @@ static auto MakeConstantForBuiltinCall(EvalContext& eval_context,
+            TypedInt, unsigned);
+        context.emitter().Emit(
+            loc_id, StringAtIndexOutOfBounds,
+            {.type = eval_context.insts().Get(arg_ids[1]).type_id(),
```
```suggestion
            {.type = eval_context.insts().Get(index_id).type_id(),
```
Should give us the same thing, right?

danakj commented on toolchain/check/eval.cpp:
```diff
@@ -1687,6 +1687,64 @@ static auto MakeConstantForBuiltinCall(EvalContext& eval_context,
+                          "index `{0}` is negative.", TypedInt);
+        context.emitter().Emit(
+            loc_id, StringIndexNegative,
+            {.type = eval_context.insts().Get(arg_ids[1]).type_id(),
```
```suggestion
            {.type = eval_context.insts().Get(index_id).type_id(),
```
Should give us the same thing right?

danakj commented on toolchain/check/eval.cpp:
```diff
@@ -1962,6 +2020,7 @@ static auto MakeConstantForCall(EvalContext& eval_context,
     // Calls to builtins might be constant.
     builtin_kind =
         eval_context.functions().Get(fn->function_id).builtin_function_kind();
+
```
```suggestion
```

danakj commented on toolchain/diagnostics/diagnostic_kind.def:
```diff
@@ -450,6 +450,8 @@ CARBON_DIAGNOSTIC_KIND(NegativeIntInUnsignedType)
 CARBON_DIAGNOSTIC_KIND(CompTimeOnlyFunctionHere)
 CARBON_DIAGNOSTIC_KIND(SelfOutsideImplicitParamList)
+CARBON_DIAGNOSTIC_KIND(StringAtIndexOutOfBounds)
+CARBON_DIAGNOSTIC_KIND(StringIndexNegative)
```
```suggestion
CARBON_DIAGNOSTIC_KIND(StringAtIndexNegative)
CARBON_DIAGNOSTIC_KIND(StringAtIndexOutOfBounds)
```

danakj commented on toolchain/check/testdata/operators/overloaded/string_indexing.carbon:
```diff
@@ -0,0 +1,299 @@
+  //@dump-sem-ir-end
+}
+
+// --- fail_wrong_type.carbon
+library "[[@TEST_NAME]]";
+
+fn TestWrongType() {
+  let x: i32 = 42;
+  //@dump-sem-ir-begin
+  // CHECK:STDERR: fail_wrong_type.carbon:[[@LINE+4]]:17: error: cannot access member of interface `Core.IndexWith(Core.IntLiteral)` in type `i32` that does not implement that interface [MissingImplInMemberAccess]
+  // CHECK:STDERR:   let c: char = x[0];
+  // CHECK:STDERR:                 ^~~~
+  // CHECK:STDERR:
+  let c: char = x[0];
+  //@dump-sem-ir-end
+}
+
```
This is not testing StringAt, it's testing that there is no `impl i32 as IndexWith`, which doesn't fit here. I think I would just drop this test. We probably don't want to test for the lack of every interface in Core for every type in Core that doesn't implement it.

What about instead testing the wrong type in the argument, like
```
class C {}
"Test"[C];
```

danakj commented on toolchain/check/testdata/operators/overloaded/string_indexing.carbon:
```diff
@@ -0,0 +1,299 @@
+
+fn TestStringIndexing() {
+  //@dump-sem-ir-begin
+  let c: char = "Test"[0];
```
Let's test boundary cases, verify that 3 works too?

danakj commented on toolchain/check/testdata/operators/overloaded/string_indexing.carbon:
```diff
@@ -0,0 +1,299 @@
+library "[[@TEST_NAME]]";
+
+fn TestLiteralNegativeIndex() {
+  //@dump-sem-ir-begin
```
For tests that are testing diagnostics, generally we can skip the semir, unless we're specifically looking for how the Errors propagate there (which I don't think we are here).

danakj reviewed (commented):
Thanks, this is looking much simpler and better now.

We could still use tests of the ValidateSignature rejecting invalid signatures. Here's some example tests to work from: https://github.com/search?q=repo%3Acarbon-language%2Fcarbon-lang+%22invalid+signature+for+builtin+function%22&type=code

ammaralassal committed:
Addressed feedback, added more tests.

danakj commented on toolchain/lower/testdata/operators/string_at.carbon:
```diff
@@ -0,0 +1,41 @@
+fn StringAt() -> char {
+    let c: char = "Test"[0];
+    return c;
+}
```
Could we also add a lower test where the StringAt operator isn't compile-time?

danakj commented on toolchain/check/testdata/operators/overloaded/string_indexing.carbon:
```diff
@@ -14,220 +14,93 @@
   //@dump-sem-ir-end
 }
 
+// --- fail_bad_decl.carbon
+library "[[@TEST_NAME]]";
+
+// CHECK:STDERR: fail_bad_decl.carbon:[[@LINE+4]]:1: error: invalid signature for builtin function "string.at" [InvalidBuiltinSignature]
+// CHECK:STDERR: fn At(s: str, index: i64) -> i32 = "string.at";
+// CHECK:STDERR: ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+// CHECK:STDERR:
+fn At(s: str, index: i64) -> i32 = "string.at";
+
+// CHECK:STDERR: fail_bad_decl.carbon:[[@LINE+4]]:1: error: invalid signature for builtin function "string.at" [InvalidBuiltinSignature]
+// CHECK:STDERR: fn At2(s: str) -> char = "string.at";
+// CHECK:STDERR: ^~~~~~~~~~~~~~~~~~~~~~~~
+// CHECK:STDERR:
+fn At2(s: str) -> char = "string.at";
+
+// CHECK:STDERR: fail_bad_decl.carbon:[[@LINE+4]]:1: error: invalid signature for builtin function "string.at" [InvalidBuiltinSignature]
+// CHECK:STDERR: fn At3(s: i32, index: i64) -> char = "string.at";
+// CHECK:STDERR: ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+// CHECK:STDERR:
+fn At3(s: i32, index: i64) -> char = "string.at";
```
We are trying to maintain a practice of putting tests in files with the smallest prelude they can have. These don't need any prelude (they can use the `none` min-prelude. So could we split them off into another file?

ammaralassal replied:
It seems that when I use `str` and `i64` in the signature, Carbon attempts to resolve those types and needs `Core`. To avoid using the full prelude in the separate test file, would a test like
```
fn At() = "string.at";
```
suffice? 

Otherwise, the test file generates the following output
```
// CHECK:STDERR: fail_bad_decl.carbon:[[@LINE+16]]:10: error: `Core.String` implicitly referenced here, but package `Core` not found [CoreNotFound]
// CHECK:STDERR: fn At(s: str, index: i64) -> i32 = "string.at";
// CHECK:STDERR:          ^~~
// CHECK:STDERR:
```

danakj replied:
Ah.. I was going to say you can use things like this in a test to avoid pulling in prelude, when you just need the types:
```
fn IntLiteral() -> type = "int_literal.make_type";
fn MakeInt(size: IntLiteral()) -> type = "int.make_type_signed";
```

But... String is written in Core in Carbon code. So we'd need to make a min_prelude for String. We can avoid that for now, then.

danakj commented on toolchain/lower/testdata/operators/string_at.carbon:
```diff
@@ -0,0 +1,41 @@
+//
+// AUTOUPDATE
+// TIP: To test this file alone, run:
+// TIP:   bazel test //toolchain/testing:file_test --test_arg=--file_tests=toolchain/lower/testdata/operators/string_at.carbon
```
nit: maybe call this file string_indexing.carbon to match the check tests?

danakj commented on toolchain/lower/testdata/operators/string_at.carbon:
```diff
@@ -0,0 +1,41 @@
+// TIP: To dump output, run:
+// TIP:   bazel run //toolchain/testing:file_test -- --dump_output --file_tests=toolchain/lower/testdata/operators/string_at.carbon
+
+fn StringAt() -> char {
```
nit: I found `StringAt` confusing at first since that's the operation we're testing inside the fn. Maybe just `F`?
```suggestion
fn F() -> char {
```

danakj reviewed (commented):
Thanks! Looks really good, a few more small things to address.

ammaralassal committed:
Addressed feedback

danakj commented on toolchain/lower/testdata/operators/string_indexing.carbon:
```diff
@@ -0,0 +1,62 @@
+}
+
+
+
```
```suggestion
```

danakj committed:
remove whitespace

danakj commented:
Just noticed you'd addressed feedback and recalled I need to poke the PR to actually make it do something, so attempting to merge, lets see if the tests pass.

danakj reviewed (approved):
LGTM

ammaralassal commented:
I suppose I have some more work to do on this, I'll go through the tester output.

ammaralassal committed:
Merge branch 'trunk' of https://github.com/ammaralassal/carbon-lang into toolchain-string-indexing
Updating branch

ammaralassal committed:
Merge branch 'toolchain-string-indexing' of https://github.com/ammaralassal/carbon-lang into toolchain-string-indexing
whitespace removal

ammaralassal committed:
Merge remote-tracking branch 'upstream/trunk' into toolchain-string-indexing
Merge upstream/trunk to get API changes

ammaralassal committed:
Addressing tester

ammaralassal committed:
Addressed tester further

ammaralassal commented:
Strange, I suppose the pre-commit failing has to do with the current issues with runners?

danakj commented:
LGTM
